File: src/main/java/xyz/devvydont/smprpg/items/blueprints/sets/fishing/WaterRod.java
package xyz.devvydont.smprpg.items.blueprints.sets.fishing;

import org.bukkit.Material;
import org.bukkit.NamespacedKey;
import org.bukkit.inventory.*;
import org.bukkit.inventory.recipe.CraftingBookCategory;
import xyz.devvydont.smprpg.SMPRPG;
import xyz.devvydont.smprpg.attribute.AttributeWrapper;
import xyz.devvydont.smprpg.items.CustomItemType;
import xyz.devvydont.smprpg.items.ItemClassification;
import xyz.devvydont.smprpg.items.attribute.AttributeEntry;
import xyz.devvydont.smprpg.items.base.CustomAttributeItem;
import xyz.devvydont.smprpg.items.blueprints.vanilla.ItemSword;
import xyz.devvydont.smprpg.items.interfaces.IBreakableEquipment;
import xyz.devvydont.smprpg.items.interfaces.ICraftable;
import xyz.devvydont.smprpg.items.interfaces.IFishingRod;
import xyz.devvydont.smprpg.services.ItemService;
import xyz.devvydont.smprpg.util.items.ToolGlobals;

import java.util.Collection;
import java.util.List;
import java.util.Set;

public class WaterRod extends CustomAttributeItem implements IBreakableEquipment, IFishingRod, ICraftable {

    public WaterRod(ItemService itemService, CustomItemType type) {
        super(itemService, type);
    }

    @Override
    public ItemClassification getItemClassification() {
        return ItemClassification.ROD;
    }

    @Override
    public Collection<AttributeEntry> getAttributeModifiers(ItemStack item) {
        return List.of(
                AttributeEntry.additive(AttributeWrapper.STRENGTH, getStrength()),
                AttributeEntry.multiplicative(AttributeWrapper.ATTACK_SPEED, -.5),
                AttributeEntry.additive(AttributeWrapper.FISHING_RATING, getFishingRating()),
                AttributeEntry.additive(AttributeWrapper.FISHING_CREATURE_CHANCE, getChance()),
                AttributeEntry.additive(AttributeWrapper.FISHING_TREASURE_CHANCE, getChance()),
                AttributeEntry.additive(AttributeWrapper.FISHING_SPEED, getSpeed())
        );
    }

    private int getSpeed() {
        return switch (this.getCustomItemType()) {
            case IRON_ROD -> 10;
            case DIAMOND_ROD -> 25;
            case PRISMARINE_ROD -> 40;
            default -> 0;
        };
    }

    @Override
    public int getPowerRating() {
        return switch (getCustomItemType()) {
            case IRON_ROD -> ToolGlobals.IRON_TOOL_POWER;
            case DIAMOND_ROD -> ToolGlobals.DIAMOND_TOOL_POWER;
            case PRISMARINE_ROD -> ToolGlobals.NETHERITE_TOOL_POWER-5;
            default -> 0;
        };
    };

    @Override
    public EquipmentSlotGroup getActiveSlot() {
        return EquipmentSlotGroup.HAND;
    }

    @Override
    public int getMaxDurability() {
        return Math.max(1, getPowerRating() * 1_000);
    }

    @Override
    public Set<FishingFlag> getFishingFlags() {
        return Set.of(FishingFlag.NORMAL);
    }

    @Override
    public NamespacedKey getRecipeKey() {
        return new NamespacedKey(SMPRPG.getPlugin(), getCustomItemType().getKey() + "_recipe");
    }

    /**
     * Work out which fishing rod this rod will be crafted from.
     */
    private RecipeChoice getTransmuteComponent() {
        return switch (this.getCustomItemType()) {
            case IRON_ROD -> new RecipeChoice.ExactChoice(ItemService.generate(Material.FISHING_ROD));
            case DIAMOND_ROD -> new RecipeChoice.ExactChoice(ItemService.generate(CustomItemType.IRON_ROD));
            case PRISMARINE_ROD -> new RecipeChoice.ExactChoice(ItemService.generate(CustomItemType.DIAMOND_ROD));
            default -> new RecipeChoice.ExactChoice(ItemService.generate(Material.BARRIER));
        };
    }

    /**
     * Get the material used for crafting the rod part of the rod.
     */
    private RecipeChoice getCraftingMaterial() {
        return switch (this.getCustomItemType()) {
            case IRON_ROD -> new RecipeChoice.ExactChoice(ItemService.generate(Material.IRON_INGOT));
            case DIAMOND_ROD -> new RecipeChoice.ExactChoice(ItemService.generate(Material.DIAMOND));
            case PRISMARINE_ROD -> new RecipeChoice.ExactChoice(ItemService.generate(CustomItemType.ENCHANTED_PRISMARINE_CRYSTAL));
            default -> new RecipeChoice.ExactChoice(ItemService.generate(Material.BARRIER));
        };
    }

    /**
     * Get the material used for crafting the rod part of the rod.
     */
    private RecipeChoice getStringMaterial() {
        return switch (this.getCustomItemType()) {
            case PRISMARINE_ROD -> new RecipeChoice.ExactChoice(ItemService.generate(CustomItemType.PREMIUM_STRING));
            default -> new RecipeChoice.ExactChoice(ItemService.generate(Material.STRING));
        };
    }

    @Override
    public CraftingRecipe getCustomRecipe() {
        var recipe = new ShapedRecipe(getRecipeKey(), generate());
        recipe.shape(
                "  m",
                " ts",
                "m s"
        );
        recipe.setIngredient('m', getCraftingMaterial());
        recipe.setIngredient('t', getTransmuteComponent());
        recipe.setIngredient('s', getStringMaterial());
        recipe.setCategory(CraftingBookCategory.EQUIPMENT);
        recipe.setGroup("water_rod");
        return recipe;
    }

    /**
     * A collection of items that will unlock the recipe for this item. Typically will be one of the components
     * of the recipe itself, but can be set to whatever is desired
     *
     * @return
     */
    @Override
    public Collection<ItemStack> unlockedBy() {
        if (getCraftingMaterial() instanceof RecipeChoice.ExactChoice exact)
            return List.of(exact.getItemStack());
        return List.of(ItemService.generate(Material.IRON_INGOT));
    }

    private int getFishingRating() {
        return switch (getCustomItemType()) {
            case IRON_ROD -> 10;
            case DIAMOND_ROD -> 25;
            case PRISMARINE_ROD -> 45;
            default -> 0;
        };
    };

    private int getStrength() {
        return (int) switch (getCustomItemType()) {
            case IRON_ROD -> ItemSword.getSwordDamage(Material.IRON_SWORD) / 2;
            case GOLD_ROD -> ItemSword.getSwordDamage(Material.GOLDEN_SWORD) / 2;
            case DIAMOND_ROD -> ItemSword.getSwordDamage(Material.DIAMOND_SWORD) / 2;
            case PRISMARINE_ROD -> ItemSword.getSwordDamage(Material.DIAMOND_SWORD) / 2 + 10;
            default -> 0;
        };
    };

    private double getChance() {
        return switch (getCustomItemType()) {
            case IRON_ROD -> 0.5;
            case DIAMOND_ROD -> 1;
            case PRISMARINE_ROD -> 2;
            default -> 0;
        };
    };
}

File: src/main/java/xyz/devvydont/smprpg/items/blueprints/sets/fishing/WaterRod.kt
package xyz.devvydont.smprpg.items.blueprints.sets.fishing

import net.kyori.adventure.key.Key
import org.bukkit.Material
import org.bukkit.inventory.*
import xyz.devvydont.smprpg.attribute.AttributeWrapper
import xyz.devvydont.smprpg.items.CustomItemType
import xyz.devvydont.smprpg.items.ItemClassification
import xyz.devvydont.smprpg.items.attribute.AttributeEntry
import xyz.devvydont.smprpg.items.attribute.AttributeEntry.Companion.additive
import xyz.devvydont.smprpg.items.attribute.AttributeEntry.Companion.multiplicative
import xyz.devvydont.smprpg.items.base.CustomAttributeItem
import xyz.devvydont.smprpg.items.blueprints.vanilla.ItemSword.Companion.getSwordDamage
import xyz.devvydont.smprpg.items.interfaces.IBreakableEquipment
import xyz.devvydont.smprpg.items.interfaces.IFishingRod
import xyz.devvydont.smprpg.items.interfaces.IFishingRod.FishingFlag
import xyz.devvydont.smprpg.items.interfaces.IModelOverridden
import xyz.devvydont.smprpg.items.interfaces.IRepairable
import xyz.devvydont.smprpg.services.ItemService
import xyz.devvydont.smprpg.util.items.ToolGlobals
import kotlin.math.max

class WaterRod(itemService: ItemService, type: CustomItemType) : CustomAttributeItem(itemService, type),
    IBreakableEquipment, IFishingRod, IModelOverridden, IRepairable {
    override val itemClassification: ItemClassification get() = ItemClassification.ROD
    override val repairMaterial: MutableCollection<ItemStack> get() = getRepairMaterialFromType()

    override fun getAttributeModifiers(item: ItemStack?): MutableCollection<AttributeEntry?> {
        return mutableListOf(
            additive(AttributeWrapper.STRENGTH, this.strength.toDouble()),
            multiplicative(AttributeWrapper.ATTACK_SPEED, -.5),
            additive(AttributeWrapper.FISHING_RATING, this.fishingRating.toDouble()),
            additive(
                AttributeWrapper.FISHING_CREATURE_CHANCE,
                this.chance
            ),
            additive(
                AttributeWrapper.FISHING_TREASURE_CHANCE,
                this.chance
            ),
            additive(AttributeWrapper.FISHING_SPEED, this.speed.toDouble())
        )
    }

    private val speed: Int
        get() = when (this.customItemType) {
            CustomItemType.IRON_ROD -> 10
            CustomItemType.MITHRIL_ROD -> 25
            CustomItemType.PRISMARINE_ROD -> 40
            else -> 0
        }

    override fun getPowerRating(): Int {
        return when (customItemType) {
            CustomItemType.IRON_ROD -> ToolGlobals.IRON_TOOL_POWER
            CustomItemType.MITHRIL_ROD -> ToolGlobals.MITHRIL_TOOL_POWER
            CustomItemType.PRISMARINE_ROD -> ToolGlobals.NETHERITE_TOOL_POWER - 5
            else -> 0
        }
    }

    override fun getActiveSlot(): EquipmentSlotGroup {
        return EquipmentSlotGroup.HAND
    }

    override fun getMaxDurability(): Int {
        return max(1, powerRating * 1000)
    }

    override fun getFishingFlags(): MutableSet<FishingFlag?> {
        return mutableSetOf(FishingFlag.NORMAL)
    }

    fun getRepairMaterialFromType() : MutableCollection<ItemStack> {
        return when (customItemType) {
            CustomItemType.IRON_ROD -> mutableListOf(itemService.getCustomItem(Material.IRON_INGOT))
            CustomItemType.MITHRIL_ROD -> mutableListOf(itemService.getCustomItem(CustomItemType.MITHRIL_INGOT))
            CustomItemType.PRISMARINE_ROD -> mutableListOf(itemService.getCustomItem(CustomItemType.ENCHANTED_PRISMARINE_CRYSTAL))
            else -> mutableListOf()
        }
    }

    private val fishingRating: Int
        get() = when (customItemType) {
            CustomItemType.IRON_ROD -> 10
            CustomItemType.MITHRIL_ROD -> 25
            CustomItemType.PRISMARINE_ROD -> 45
            else -> 0
        }

    private val strength: Int
        get() = when (customItemType) {
            CustomItemType.IRON_ROD -> getSwordDamage(Material.IRON_SWORD) / 2
            CustomItemType.GOLD_ROD -> getSwordDamage(Material.GOLDEN_SWORD) / 2
            CustomItemType.MITHRIL_ROD -> getSwordDamage(CustomItemType.TITANIUM_SWORD) / 2
            CustomItemType.PRISMARINE_ROD -> getSwordDamage(Material.DIAMOND_SWORD) / 2 + 10
            else -> 0
        }.toInt()

    private val chance: Double
        get() = when (customItemType) {
            CustomItemType.IRON_ROD -> 0.5
            CustomItemType.MITHRIL_ROD -> 1
            CustomItemType.PRISMARINE_ROD -> 2
            else -> 0
        }.toDouble()

    override fun getDisplayKey(): Key { return IModelOverridden.ofItemType(customItemType) }
}
